dds/DCPS/transport/rtps_udp/RtpsUdpSendStrategy.cpp

```diff
@@ -319,6 +343,52 @@ RtpsUdpSendStrategy::encode_payload(const RepoId& pub_id,
   }
 }
 
+ACE_Message_Block*
+RtpsUdpSendStrategy::pre_send_packet(const ACE_Message_Block* plain)
+{
+  const DDS::Security::CryptoTransform_var crypto = link_->security_config()->get_crypto_transform();
+  if (!crypto) {
+    return plain->duplicate();
+  }
+
+  bool stateless_or_volatile = false;
+  Message_Block_Ptr submessages(encode_submessages(plain, crypto, stateless_or_volatile));
+
+  if (!submessages || stateless_or_volatile || link_->local_crypto_handle() == DDS::HANDLE_NIL) {
+    return submessages.release();
+  }
+
+  return encode_rtps_message(submessages.get(), crypto);
+}
+
+ACE_Message_Block*
+RtpsUdpSendStrategy::encode_rtps_message(const ACE_Message_Block* plain, DDS::Security::CryptoTransform* crypto)
+{
+  using namespace DDS::Security;
+  DDS::OctetSeq encoded_rtps_message;
+  const DDS::OctetSeq plain_rtps_message = toSeq(plain);
+  const ParticipantCryptoHandle send_handle = link_->local_crypto_handle();
+  const ParticipantCryptoHandleSeq recv_handles; // unused
+  int idx = 0; // unused
+  SecurityException ex = {"", 0, 0};
+  if (crypto->encode_rtps_message(encoded_rtps_message, plain_rtps_message,
+                                  send_handle, recv_handles, idx, ex)) {
+    Message_Block_Ptr out(new ACE_Message_Block(encoded_rtps_message.length()));
+    const char* raw = reinterpret_cast<const char*>(encoded_rtps_message.get_buffer());
+    out->copy(raw, encoded_rtps_message.length());
+    return out.release();
+  }
+  if (ex.code == 0 && ex.minor_code == 0) {
```

Comment:
Should we make a shorthand method for checking if the security exception hasn't been set? And does SecurityException initialize to undefined or a non-error state?

Reply:
The spec says:
```
This operation may optionally not perform any transformation of the input RTPS message. In this case,
the operation shall return false but not set the exception object. In this situation the DDS
implementation shall send the original RTPS message.
```

If this happened in multiple places I'd be for creating a more systematic way of handing it, but it's only this one case.